### PR TITLE
Annotate SQLGlot with python types (WIP)

### PR DESCRIPTION
**Pull request by GeorgeSittas**
Opening this PR to start a discussion about adding type annotations throughout SQLGlot. I don't have a lot of experience with typing in python, so there may be some errors on the way. Any feedback is greatly appreciated!

As an extra, I'll try to polish/add documentation and maybe refactor some bits of code that can be improved.

### CODE REVIEW

File: sqlglot/__init__.py [hunk 1]
```diff
@@ -33,38 +35,36 @@
 schema = MappingSchema()
 
 
-def parse(sql, read=None, **opts):
+def parse(sql: str, read: t.Optional[str] = None, **opts) -> t.List[t.Optional[Expression]]:
```
Comment (barakalon) [inline sqlglot/__init__.py:38]: `Optional[Expression]` means `Expression | None`. Can elements of the list be None?

File: sqlglot/__init__.py [hunk 2]
```diff
@@ -33,38 +35,36 @@
...
+        The resulting syntax tree collection.
     """
     dialect = Dialect.get_or_raise(read)()
     return dialect.parse(sql, **opts)
 
 
-def parse_one(sql, read=None, into=None, **opts):
+def parse_one(
+    sql: str, read: t.Optional[str] = None, into: t.Optional[Expression] = None, **opts
+) -> t.Optional[Expression]:
```
Comment (barakalon) [inline sqlglot/__init__.py:56]: Same question here - can this return None?

File: sqlglot/__init__.py [hunk 3]
```diff
@@ -33,38 +35,36 @@
...
-        typing.List[Expression]: the list of parsed syntax trees.
+        The resulting syntax tree collection.
     """
     dialect = Dialect.get_or_raise(read)()
     return dialect.parse(sql, **opts)
 
 
-def parse_one(sql, read=None, into=None, **opts):
+def parse_one(
+    sql: str, read: t.Optional[str] = None, into: t.Optional[Expression] = None, **opts
```
Comment (tobymao) [inline sqlglot/__init__.py:55]: into can also be a string, so you need a union

t.Optional[Expression | str] 

and add the future annotations at the top to support this

**Comment by tobymao** [comment]
you need to turn on mypy for these files so it actually checks it

File: sqlglot/errors.py [hunk 4]
```diff
@@ -30,7 +31,7 @@ class OptimizeError(SqlglotError):
     pass
 
 
-def concat_errors(errors, maximum):
+def concat_errors(errors: t.Union[ErrorLevel, str], maximum: int) -> str:
```
Comment (barakalon) [inline sqlglot/errors.py:34]: I don't think `Union[ErrorLevel, str]` is right. I think it should be `Sequence[Any]`

File: sqlglot/helper.py [hunk 5]
```diff
@@ -7,40 +7,57 @@
 from copy import copy
 from enum import Enum
 
+from sqlglot.expressions import Expression, Table
```
Comment (tobymao) [inline sqlglot/helper.py:10]: if t.TYPE_CHECKING:

File: sqlglot/helper.py [hunk 6]
```diff
@@ -7,40 +7,57 @@
...
+def list_get(arr: t.List[t.Any], index: int) -> t.Optional[t.Any]:
+    """Returns the value in `arr` at position `index`, or `None` if `index` is out of bounds."""
     try:
         return arr[index]
     except IndexError:
         return None
 
 
-def ensure_list(value):
+def ensure_list(value: t.Any) -> t.Union[t.List[t.Any], t.Tuple[t.Any], t.Set[t.Any]]:
```
Comment (tobymao) [inline sqlglot/helper.py:29]: let's | instead of union

File: sqlglot/helper.py [hunk 7]
```diff
@@ -7,40 +7,57 @@
...
 
 
 class AutoName(Enum):
-    def _generate_next_value_(name, _start, _count, _last_values):
+    def _generate_next_value_(name: str, _start: int, _count: int, _last_values: int) -> str:
         return name
 
 
-def list_get(arr, index):
+def list_get(arr: t.List[t.Any], index: int) -> t.Optional[t.Any]:
```
Comment (barakalon) [inline sqlglot/helper.py:21]: Use a TypeVar:

```python
T = t.TypeVar("T")

def list_get(arr: t.Sequence[T], index: int) -> t.Optional[T]:
    ...
```

File: sqlglot/helper.py (hunk 6 above)
Comment (barakalon) [inline sqlglot/helper.py:29]: The proper thing to use here is [overload](https://docs.python.org/3/library/typing.html#typing.overload).

File: sqlglot/helper.py [hunk 8]
```diff
@@ -53,7 +70,18 @@ def subclasses(module_name, classes, exclude=()):
     ]
 
 
-def apply_index_offset(expressions, offset):
+def apply_index_offset(expressions: t.List[Expression], offset: int) -> t.List[Expression]:
```
Comment (barakalon) [inline sqlglot/helper.py:73]: This should also use a TypeVar

File: sqlglot/helper.py (hunk 6 above)
Comment (eakmanrq) [inline sqlglot/helper.py:29]: The fact that this doesn't actually ensure a list type becomes even more confusing when we start adding types. Since we don't like the cost of casting to list can we rename to `ensure_iterable`? Also should we establish a term in the codebase for referencing an iterable that is not a str or byte and then use that in the name? Maybe `ensure_collection`?

File: sqlglot/helper.py [hunk 9]
```diff
@@ -64,14 +92,26 @@ def apply_index_offset(expressions, offset):
...
 
 
-def camel_to_snake_case(name):
+def camel_to_snake_case(name: str) -> str:
+    """Converts `name` from camelCase to snake_case and returns the result."""
     return CAMEL_CASE_PATTERN.sub("_", name).upper()
 
 
-def while_changing(expression, func):
+def while_changing(expression: t.Optional[Expression], func: t.Callable) -> t.Optional[Expression]:
```
Comment (barakalon) [inline sqlglot/helper.py:104]: I think this is what you want:

```python
E = t.TypeVar("E", bound=Expression)

def while_changing(expression: t.Optional[Expression], func: t.Callable[[Expression], E]) -> E:
```

The TypeVar might be overkill. But in any case, I think we should annotate the Callable.

File: sqlglot/helper.py [hunk 10]
```diff
@@ -80,10 +120,19 @@ def while_changing(expression, func):
     return expression
 
 
-def tsort(dag):
+def tsort(dag: t.Dict[t.Any, t.List[t.Any]]) -> t.List[t.Any]:
```
Comment (barakalon) [inline sqlglot/helper.py:123]: Use a TypeVar

File: sqlglot/helper.py (hunk 6 above)
Comment (tobymao) [reply sqlglot/helper.py:29]: ensure_collection is fine

Comment (GeorgeSittas) [reply sqlglot/helper.py:29]: Agreed -- I had some trouble deciding on a name for this function. I wasn't sure what was the intent we wanted to capture because there are generally two uses throughout the codebase:

1. Convert something into a list, because we expect it to be a list in other parts of the code.
2. Convert something into a list only so that we can iterate over it.

If we do `ensure_iterable`, it might obscure the intent for case 1.

File: sqlglot/__init__.py (hunk 1 above)
Comment (GeorgeSittas) [reply sqlglot/__init__.py:38]: Yes, for instance:
```python
>>> import sqlglot
>>> sqlglot.parse("")
[None]
```

File: sqlglot/helper.py (hunk 6 above)
Comment (eakmanrq) [reply sqlglot/helper.py:29]: This might be a pain but my vote would be to have `ensure_list` and `ensure_collection` and then functions would use the one appropriate for their use case. For example in the DataFrame implementation I really do want a list so I would use `ensure_list` (and take the performance consequences) while other places we might actually be fine with a collection.

**Comment by GeorgeSittas** [comment]
> you need to turn on mypy for these files so it actually checks it

I think we're already using mypy, no? This is in `run_checks.sh` currently:
```bash
python -m mypy sqlglot tests
```

**Comment by tobymao** [comment]
> > you need to turn on mypy for these files so it actually checks it
> 
> I think we're already using mypy, no? This is in `run_checks.sh` currently:
> 
> ```shell
> python -m mypy sqlglot tests
> ```

check setup.cfg, i think it ignores all erros in sqlglot except datafram.*

File: sqlglot/helper.py (hunk 6 above)
Comment (GeorgeSittas) [reply sqlglot/helper.py:29]: Got it, I'll try that.

File: sqlglot/errors.py (hunk 4 above)
Comment (GeorgeSittas) [reply sqlglot/errors.py:34]: Oops, ErrorLevel is definitely wrong there. Although does `Sequence` capture exceptions? For example, in the parser we call `concat_errors` with `self.errors`, which can have a type of `List[ParseError]`.

Comment (barakalon) [reply sqlglot/errors.py:34]: `Sequence` is an abstract type. I think it works for both List and Tuple? I might have that wrong...

But in either case, the element type is Any. We simply call `str` on the elements, which I think works for all objects.

Comment (GeorgeSittas) [reply sqlglot/errors.py:34]: Yeah you're right, I got confused. Will fix this too.

**Comment by GeorgeSittas** [comment]
I tried to address all raised issues:

- @barakalon I was skeptical regarding `ensure_list` and `ensure_collection` types.. Is the `@overload` used correctly?
- AFAIK python doesn't compute the `for` loop's iterable >1 times, so I think it should be ok to simplify the related parts.

I activated mypy for the entire sqlglot module, so as expected it yells a lot right now. I'll try to fix them up one by one by annotating the rest of the code soon.

File: sqlglot/helper.py [hunk 11]
```diff
@@ -187,16 +286,38 @@ def split_num_words(value: str, sep: str, min_num_words: int, fill_from_start: b
...
+        >>> is_collection("test")
+        False
+
+    Args:
+        value: the value to check if it is a collection.
+
+    Returns:
+        A `bool` value indicating if it is a collection.
+    """
+    return issubclass(type(value), Collection) and not isinstance(value, (str, bytes))
```
Comment (barakalon) [inline sqlglot/helper.py:315]: Can this not be `isinstance(value, Collection)`?

**Review by barakalon — COMMENTED** [review]
Looks right to me

**Comment by GeorgeSittas** [comment]
A question I had while annotating `schema.py` was whether it's correct to use `type: ignore`, for example in `column_names`. The reason I did that is because there were some errors when I did `table.this` etc., due to the `table` parameter having a type of `exp.Table | str`, even though we know it's been converted to `Table`. Does this make sense?

File: sqlglot/__init__.py [hunk 12]
```diff
@@ -33,38 +37,36 @@
 schema = MappingSchema()
 
 
-def parse(sql, read=None, **opts):
+def parse(sql: str, read: t.Optional[str] = None, **opts) -> t.List[t.Optional[Expression]]:
```
Comment (tobymao) [inline sqlglot/__init__.py:40]: read can also be a dialect enum i think so str | Dialect

File: sqlglot/__init__.py [hunk 13]
```diff
@@ -33,38 +37,36 @@
...
-        typing.List[Expression]: the list of parsed syntax trees.
+        The resulting syntax tree collection.
     """
     dialect = Dialect.get_or_raise(read)()
     return dialect.parse(sql, **opts)
 
 
-def parse_one(sql, read=None, into=None, **opts):
+def parse_one(
+    sql: str, read: t.Optional[str] = None, into: t.Optional[Expression | str] = None, **opts
```
Comment (tobymao) [inline sqlglot/__init__.py:57]: same here

File: sqlglot/__init__.py [hunk 14]
```diff
@@ -77,25 +79,29 @@ def parse_one(sql, read=None, into=None, **opts):
     return result[0] if result else None
 
 
-def transpile(sql, read=None, write=None, identity=True, error_level=None, **opts):
+def transpile(
+    sql: str,
+    read: t.Optional[str] = None,
```
Comment (tobymao) [inline sqlglot/__init__.py:84]: same

File: sqlglot/helper.py [hunk 15]
```diff
@@ -1,48 +1,112 @@
...
 class AutoName(Enum):
-    def _generate_next_value_(name, _start, _count, _last_values):
+    """This is used for creating enum classes where `auto()` is the string form of the corresponding value's name."""
+
+    def _generate_next_value_(name, _start, _count, _last_values):  # type: ignore
         return name
 
 
-def list_get(arr, index):
+def sequence_get(seq: t.Sequence[T], index: int) -> t.Optional[T]:
```
Comment (tobymao) [inline sqlglot/helper.py:30]: maybe just seq_get

File: sqlglot/helper.py [hunk 16]
```diff
@@ -1,48 +1,112 @@
...
         return []
-    return value if isinstance(value, (list, tuple, set)) else [value]
+    elif isinstance(value, (list, tuple)):
+        return list(value)
 
+    return [value]
+
+
+@t.overload
+def ensure_collection(value: t.Collection[T]) -> t.Collection[T]:
```
Comment (tobymao) [inline sqlglot/helper.py:57]: this is how you do this?

File: sqlglot/helper.py [hunk 17]
```diff
@@ -1,48 +1,112 @@
...
+    Args:
+        value: the value of interest.
 
-def csv(*args, sep=", "):
+    Returns:
+        The value if it's a collection, or else the value wrapped in a list.
+    """
+    if value is None:
+        return []
+    return value if is_collection(value) else [value]
```
Comment (tobymao) [inline sqlglot/helper.py:78]: is_collection doesn't need to be another function call, you don't use it anywhere

File: sqlglot/helper.py [hunk 18]
```diff
@@ -1,48 +1,112 @@
...
+def ensure_collection(value: t.Collection[T]) -> t.Collection[T]:
+    ...
+
+
+@t.overload
+def ensure_collection(value: T) -> t.Collection[T]:
+    ...
+
+
+def ensure_collection(value):
```
Comment (tobymao) [inline sqlglot/helper.py:66]: can't this just be 

```
def ensure_collection(value: T | t.Collection[T]) -> t.Collection[T] 
```

or does the type checker get confused?

File: sqlglot/helper.py [hunk 19]
```diff
@@ -53,7 +117,18 @@ def subclasses(module_name, classes, exclude=()):
     ]
 
 
-def apply_index_offset(expressions, offset):
+def apply_index_offset(expressions: t.List[E], offset: int) -> t.List[E]:
```
Comment (tobymao) [inline sqlglot/helper.py:120]: why E? why not just use Expression you have it in t.TYpeChecking now

File: sqlglot/helper.py [hunk 20]
```diff
@@ -119,14 +213,14 @@ def open_file(file_name):
 
 
 @contextmanager
-def csv_reader(table):
+def csv_reader(table: Table) -> t.Any:
```
Comment (tobymao) [inline sqlglot/helper.py:216]: returns a csv reader then?

File: sqlglot/schema.py [hunk 21]
```diff
@@ -60,57 +71,62 @@ class MappingSchema(Schema):
...
 
         Args:
-            table (sqlglot.expressions.Table|str): Table expression instance or string representing the table
-            column_mapping (dict|str|sqlglot.dataframe.sql.types.StructType|list): A column mapping that describes the structure of the table
+            table: the `Table` expression instance or string representing the table.
+            column_mapping: a column mapping that describes the structure of the table.
         """
-        table = exp.to_table(table)
-        self._validate_table(table)
+        table = exp.to_table(table)  # type: ignore
```
Comment (tobymao) [inline sqlglot/schema.py:104]: why do you need ignore here

File: sqlglot/helper.py (hunk 16 above)
Comment (GeorgeSittas) [reply sqlglot/helper.py:57]: I believe so yes; if I'm not mistaken that's what Barak suggested a few days ago too.

File: sqlglot/helper.py (hunk 18 above)
Comment (GeorgeSittas) [reply sqlglot/helper.py:66]: > or does the type checker get confused?

That seems to be the case:
```python
sqlglot/helper.py:88: error: List item 0 has incompatible type "Union[T, str, bytes]"; expected "T"
```

File: sqlglot/helper.py (hunk 19 above)
Comment (barakalon) [reply sqlglot/helper.py:120]: I recommended this.
```python
l = [exp.Table(...), exp.Table(...)  # this is list[exp.Table]
apply_index_offset(l)  # using TypeVar, the return type is exp.Table. Without it, it's exp.Expression
```

I know, I know. You love type annotations Toby.

File: sqlglot/helper.py (hunk 18 above)
Comment (barakalon) [reply sqlglot/helper.py:66]: If value is `T | t.Collection[T]`, `T` will swallow up any type.
```python
ensure_collection([])  # the return type would be t.Collection[list]
```

File: sqlglot/helper.py (hunk 19 above)
Comment (tobymao) [reply sqlglot/helper.py:120]: ah i see

File: sqlglot/helper.py (hunk 18 above)
Comment (GeorgeSittas) [reply sqlglot/helper.py:66]: Right, whereas using overloads we expect mypy to explore all variants in the same order as they're declared, correct?

Comment (barakalon) [reply sqlglot/helper.py:66]: I think so. Might want to verify that.

Comment (GeorgeSittas) [reply sqlglot/helper.py:66]: Check, see here: https://mypy.readthedocs.io/en/stable/more_types.html#type-checking-calls-to-overloads

TL;DR
```
When you call an overloaded function, mypy will infer the correct return type by picking the best matching variant, after taking into consideration both the argument types and arity. However, a call is never type checked against the implementation. This is why mypy will report calls like mouse_event(5, 25, 3) as being invalid even though it matches the implementation signature.

If there are multiple equally good matching variants, mypy will select the variant that was defined first. [...]
```

File: sqlglot/helper.py (hunk 20 above)
Comment (GeorgeSittas) [reply sqlglot/helper.py:216]: I did some investigation about the csv module, and I think it doesn't expose its internal reader/writer classes. That's why I used `t.Any`. For example, I think the problem is made clear in [this thread](https://stackoverflow.com/questions/51264355/how-to-type-annotate-object-returned-by-csv-writer).

File: sqlglot/schema.py (hunk 21 above)
Comment (GeorgeSittas) [reply sqlglot/schema.py:104]: Discussed in slack.

**Review by tobymao — APPROVED** [review]
